File: android/app/src/main/java/com/metanav/app/frames/FrameSource.kt
package com.metanav.app.frames

import android.graphics.Bitmap
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.StateFlow

/** One RGB frame ready for the vision models. Portrait, origin top-left. */
class RgbFrame(val bitmap: Bitmap, val timestampMs: Long)

enum class SourceKind { GLASSES, PHONE_CAMERA }

sealed class SourceState {
    data object Idle : SourceState()
    data object Connecting : SourceState()
    data object WaitingForDevice : SourceState()
    data object Streaming : SourceState()
    data object Paused : SourceState()
    data class Error(val message: String) : SourceState()
}

/** Anything that can deliver camera frames: the glasses over DAT, or the phone camera. */
interface FrameSource {
    val kind: SourceKind
    val state: StateFlow<SourceState>
    val frames: Flow<RgbFrame>
    suspend fun start()
    fun stop()
}

File: ios/Metanav/Frames/FrameSource.swift
import CoreGraphics
import CoreVideo
import Foundation

/// A frame as delivered by a camera. Raw glasses frames and phone frames arrive as pixel buffers;
/// a decoded fallback path can hand over a CGImage instead. Portrait, origin top-left.
enum FrameImage {
    case pixelBuffer(CVPixelBuffer)
    case cgImage(CGImage)

    var width: Int {
        switch self {
        case .pixelBuffer(let pb): return CVPixelBufferGetWidth(pb)
        case .cgImage(let img): return img.width
        }
    }

    var height: Int {
        switch self {
        case .pixelBuffer(let pb): return CVPixelBufferGetHeight(pb)
        case .cgImage(let img): return img.height
        }
    }
}

struct CapturedFrame {
    let image: FrameImage
    let timestampMs: Int64
}

enum SourceKind: String, Codable, CaseIterable {
    case glasses
    case phoneCamera
}

enum SourceState: Equatable {
    case idle
    case connecting
    case waitingForDevice
    case streaming
    case paused
    case error(String)
}

/// Anything that can deliver camera frames: the glasses over the DAT SDK, or the phone camera.
protocol FrameSource: AnyObject {
    var kind: SourceKind { get }
    /// Called on arbitrary threads.
    var onState: ((SourceState) -> Void)? { get set }
    /// Called on arbitrary threads at camera rate; keep it cheap and never block.
    var onFrame: ((CapturedFrame) -> Void)? { get set }
    func start() async
    func stop()
}

func nowMs() -> Int64 { Int64(Date().timeIntervalSince1970 * 1000) }
